### PR TITLE
[Ray Air] followup to the PR (#26493) set the correct gpu id in TorchTrainer (2.0)

### PR DESCRIPTION
**Pull request by JiahaoYao**
This PR ticket is to address the followups in the pr (https://github.com/ray-project/ray/pull/26493)

- scope this down to ray_2_node_2_gpu for simplicity.
- update the error message to be more actionable: such as asking the user to post an issue on Github

cc @amogkam @richardliaw 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by JiahaoYao** [comment]
fix this issue 

```python


=================================== FAILURES ===================================
--
  | _____________________ test_tune_torch_get_device_gpu[0.5] ______________________
  |  
  | ray_2_node_4_gpu = None, num_gpus_per_worker = 0.5
  |  
  | @pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2])
  | def test_tune_torch_get_device_gpu(ray_2_node_4_gpu, num_gpus_per_worker):
  | from ray import tune
  | from ray.tune.tuner import Tuner, TuneConfig
  |  
  | num_samples = 2
  |  
  | @patch("torch.cuda.is_available", lambda: True)
  | def train_func():
  | train.report(device_id=train.torch.get_device().index)
  |  
  | trainer = TorchTrainer(
  | train_func,
  | torch_config=TorchConfig(backend="gloo"),
  | scaling_config={
  | "num_workers": 2,
  | "use_gpu": True,
  | >               "resources_per_worker": {"GPU": num_gpus_per_worker},
  | },
  | )
  |  
  | python/ray/train/tests/test_tune.py:341:
  | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  | /ray/python/ray/train/torch/torch_trainer.py:203: in __init__
  | resume_from_checkpoint=resume_from_checkpoint,
  | /ray/python/ray/train/data_parallel_trainer.py:269: in __init__
  | resume_from_checkpoint=resume_from_checkpoint,
  | /ray/python/ray/train/base_trainer.py:161: in __init__
  | self._validate_attributes()
  | /ray/python/ray/train/data_parallel_trainer.py:273: in _validate_attributes
  | super()._validate_attributes()
  | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  |  
  | self = <ray.train.torch.torch_trainer.TorchTrainer object at 0x7f51ec98b710>
  |  
  | def _validate_attributes(self):
  | """Called on __init()__ to validate trainer attributes."""
  | # Run config
  | if not isinstance(self.run_config, RunConfig):
  | raise ValueError(
  | f"`run_config` should be an instance of `ray.air.RunConfig`, "
  | f"found {type(self.run_config)} with value `{self.run_config}`."
  | )
  | # Scaling config
  | if not isinstance(self.scaling_config, ScalingConfig):
  | raise ValueError(
  | >               "`scaling_config` should be an instance of `ScalingConfig`, "
  | f"found {type(self.scaling_config)} with value `{self.scaling_config}`."
  | )
  | E           ValueError: `scaling_config` should be an instance of `ScalingConfig`, found <class 'dict'> with value `{'num_workers': 2, 'use_gpu': True, 'resources_per_worker': {'GPU': 0.5}}`.
  |  


```

File: python/ray/train/torch/train_loop_utils.py [hunk 1]
```diff
@@ -503,7 +503,8 @@ def get_device(self) -> torch.device:
                     device_id = cuda_visible_list.index(gpu_id)
                 else:
                     raise RuntimeError(
-                        f"CUDA_VISIBLE_DEVICES set incorrectly: {cuda_visible_str}"
+                        f"CUDA_VISIBLE_DEVICES set incorrectly: {cuda_visible_str}."
+                        "Please submit an issue on GitHub!"
```
Comment (richardliaw) [inline python/ray/train/torch/train_loop_utils.py:507]:
```suggestion
                        f"CUDA_VISIBLE_DEVICES set incorrectly. Got {cuda_visible_str}, expected to include {gpu_id}"
                        "Did you override CUDA_VISIBLE_DEVICES? If not, please help file an issue on Github."
```

File: python/ray/train/tests/test_tune.py [hunk 2]
```diff
@@ -325,21 +325,24 @@ def train_func():
...
+
     trainer = TorchTrainer(
         train_func,
         torch_config=TorchConfig(backend="gloo"),
-        scaling_config={
-            "num_workers": 2,
-            "use_gpu": True,
-            "resources_per_worker": {"GPU": num_gpus_per_worker},
-        },
+        scaling_config=ScalingConfig(*scaling_config),
```
Comment (richardliaw) [inline python/ray/train/tests/test_tune.py:345]: this change got merged in in the morning, can you pull in master?

File: python/ray/train/tests/test_gpu.py [hunk 3]
```diff
@@ -111,14 +111,13 @@ def train_fn():
 
     count = Counter(devices)
     if num_gpus_per_worker == 0.5:
-        for i in range(4):
+        for i in range(2):
             assert count[i] == 4
```
Comment (richardliaw) [inline python/ray/train/tests/test_gpu.py:115]: can you add a comment in the code about the expected setup (workers <> gpus topology)?

File: python/ray/train/tests/test_gpu.py [hunk 4]
```diff
@@ -111,14 +111,13 @@ def train_fn():
...
     elif num_gpus_per_worker == 1:
-        for i in range(4):
+        for i in range(2):
             assert count[i] == 2
     elif num_gpus_per_worker == 2:
-        for i in range(2):
-            assert count[2 * i] == 2
+        assert count[0] == 2
```
Comment (richardliaw) [inline python/ray/train/tests/test_gpu.py:120]: @amogkam  had a separate concern about the other test that you wrote -- https://github.com/ray-project/ray/pull/26493/files#r924064169

could you address that in this PR?

**Review by richardliaw — CHANGES_REQUESTED** [review]
thanks for the fast turnaround! just a couple requests.

File: python/ray/train/tests/test_tune.py (hunk 2 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_tune.py:345]: done.

File: python/ray/train/tests/test_gpu.py (hunk 3 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_gpu.py:122]: done

File: python/ray/train/tests/test_gpu.py (hunk 4 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_gpu.py:134]: > Hmm what exactly is this test supposed to be testing and how reliable is it? If we have 2 trials, each with 1 gpu (2*0.5) then the placement of the trials can be on either node, and we will get different results each time?

Example: in this case, if there is no preference of using the node for the ray. i agree with @amogkam , the results would be different. 
> node1 [0, 1], node2 [0, 1]

If the two trials are on the same node, it would give `[0, 0, 1, 1]`. 
If the two trials are on the different nodes, it would give `[0, 0, 0, 0]`. 

---------
The good thing about this test, `num_workers=int(4 / num_gpus_per_worker)` we always saturate the use of the gpus on the node. In this setup, the output might not be varied if there are launched at the same time.

**Review by JiahaoYao — COMMENTED** [review]
update my review

File: python/ray/train/tests/test_gpu.py [hunk 5]
```diff
@@ -110,15 +111,27 @@ def train_fn():
     trainer.shutdown()
 
     count = Counter(devices)
+    # cluster setups: 2 nodes, 2 gpus per node
+    # `CUDA_VISIBLE_DEVICES` is set to "0,1" on node 1 and node 2
     if num_gpus_per_worker == 0.5:
-        for i in range(4):
+        # worker gpu topology:
+        # 4 workers on node 1, 4 workers on node 2
+        # `ray.get_gpu_ids()` returns [0], [0], [1], [1] on node 1
```
Comment (amogkam) [inline python/ray/train/tests/test_gpu.py:119]: Nice, thanks for adding these comments!

File: python/ray/train/tests/test_gpu.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/tests/test_gpu.py:134]: Right for this test we are fully saturating the GPUs, but I'm referring to the test in `test_tune`. 

For that test we've hardcoded `nums_samples=2` and `num_workers: 2`. 

Also, what is that test testing for? Is it needed if we have this test in `test_gpu` already?

File: python/ray/train/tests/test_gpu.py (hunk 5 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_gpu.py:119]: :cool:

**Review by JiahaoYao — COMMENTED** [review]
> Right for this test we are fully saturating the GPUs, but I'm referring to the test in test_tune.

> For that test we've hardcoded nums_samples=2 and num_workers: 2.

> Also, what is that test testing for? Is it needed if we have this test in test_gpu already?


got it, thanks @amogkam! the one in the `test_tune` is a test for the ray.tune environment. Since the ray.tune, it separates the environment. since the fix for the GPU is for the local rank, actually ray.tune environment is more important testcase (ray.tune will separate the gpu environment, but ray trainer will always share the gpu environment, I mean the local gpu id is global gpu id.) 

for the ray.tune, even the actor is on the gpu id (1), the local rank will still go back to 0. so, the device id is 0. thus, no ambiguous results for the ray.tune.


-------
TL;DR
- ray.tune is actually the real test we want. 
- ray.tune will separate the environment and the new fix will always point to the local rank. 
- whether the second trials run on the first node or second node does not matter because the local rank for the GPU is always 0. 
- also happy to change the `test_tune` to better test!

**Comment by amogkam** [comment]
Discussed offline- we can make this test more generic rather than specifically for Tune, and also make sure all the device ids are set correctly.

```
@pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2], expected_result=[[0, 1], [0, 1, 2]])
def test_torch_device_nested_actor(ray_2_node_4_gpu, num_gpus_per_worker, 
                                   expected_result):
    """Tests if GPU ids are set correctly when running train concurrently in nested 
    actors (for example when used with Tune)."""
    from ray import tune
    from ray.tune.tuner import Tuner, TuneConfig

    num_samples = 2
    num_workers = 2

    @patch("torch.cuda.is_available", lambda: True)
    def train_func():
        return train.torch.get_device().index

    trainer = TorchTrainer(
        train_func,
        torch_config=TorchConfig(backend="gloo"),
        scaling_config=ScalingConfig(
            num_workers=num_workers,
            use_gpu=True,
            resources_per_worker={"GPU": num_gpus_per_worker},
        ),
    )
    
    @ray.remote(num_gpus=2)
    class TrialActor:
        def __init__(self, trainer):
            self.trainer = trainer
        
        def run(self):
            ids = self.trainer.train()
            return ids
        
    actors = [TrialActor.remote(trainer) for _ in range(num_samples)]
    device_ids = [actor.run.remote() for actor in actors]

    assert len(device_ids) == num_samples
    for i in range(num_samples):
        for j in range(num_workers):
            assert device_ids[i][j] == 0
```

**Comment by JiahaoYao** [comment]
```python
object_refs = [ObjectRef(32d950ec0ccf9d2a419008bb7d515502da03c5d60100000001000000)]

    @PublicAPI
    @client_mode_hook(auto_init=True)
    def get(
        object_refs: Union[ray.ObjectRef, Sequence[ray.ObjectRef]],
        *,
        timeout: Optional[float] = None,
    ) -> Union[Any, List[Any]]:
        """Get a remote object or a list of remote objects from the object store.
    
        This method blocks until the object corresponding to the object ref is
        available in the local object store. If this object is not in the local
        object store, it will be shipped from an object store that has it (once the
        object has been created). If object_refs is a list, then the objects
        corresponding to each object in the list will be returned.
    
        Ordering for an input list of object refs is preserved for each object
        returned. That is, if an object ref to A precedes an object ref to B in the
        input list, then A will precede B in the returned list.
    
        This method will issue a warning if it's running inside async context,
        you can use ``await object_ref`` instead of ``ray.get(object_ref)``. For
        a list of object refs, you can use ``await asyncio.gather(*object_refs)``.
    
        Args:
            object_refs: Object ref of the object to get or a list of object refs
                to get.
            timeout (Optional[float]): The maximum amount of time in seconds to
                wait before returning.
    
        Returns:
            A Python object or a list of Python objects.
    
        Raises:
            GetTimeoutError: A GetTimeoutError is raised if a timeout is set and
                the get takes longer than timeout to return.
            Exception: An exception is raised if the task that created the object
                or that created one of the objects raised an exception.
        """
        worker = global_worker
        worker.check_connected()
    
        if hasattr(worker, "core_worker") and worker.core_worker.current_actor_is_asyncio():
            global blocking_get_inside_async_warned
            if not blocking_get_inside_async_warned:
                logger.warning(
                    "Using blocking ray.get inside async actor. "
                    "This blocks the event loop. Please use `await` "
                    "on object ref with asyncio.gather if you want to "
                    "yield execution to the event loop instead."
                )
                blocking_get_inside_async_warned = True
    
        with profiling.profile("ray.get"):
            is_individual_id = isinstance(object_refs, ray.ObjectRef)
            if is_individual_id:
                object_refs = [object_refs]
    
            if not isinstance(object_refs, list):
                raise ValueError(
                    "'object_refs' must either be an object ref "
                    "or a list of object refs."
                )
    
            # TODO(ujvl): Consider how to allow user to retrieve the ready objects.
            values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
            for i, value in enumerate(values):
                if isinstance(value, RayError):
                    if isinstance(value, ray.exceptions.ObjectLostError):
                        worker.core_worker.dump_object_store_memory_usage()
                    if isinstance(value, RayTaskError):
                        raise value.as_instanceof_cause()
                    else:
>                       raise value
E                       ray.exceptions.RayActorError: The actor died because of an error raised in its creation task, ray::TrialActor.__init__() (pid=72220, ip=127.0.0.1, repr=<pyt2.TrialActor object at 0x7fc80484ce90>)
E                         At least one of the input arguments for this task could not be computed:
E                       ray.exceptions.RaySystemError: System error: maximum recursion depth exceeded while calling a Python object
E                       traceback: Traceback (most recent call last):
E                         File "/Users/jimmy/opt/anaconda3/envs/dl/lib/python3.7/site-packages/ray/train/_internal/utils.py", line 183, in __getattr__
E                           actor_method = getattr(self.actor, item)
E                         File "/Users/jimmy/opt/anaconda3/envs/dl/lib/python3.7/site-packages/ray/train/_internal/utils.py", line 183, in __getattr__
E                           actor_method = getattr(self.actor, item)
E                         File "/Users/jimmy/opt/anaconda3/envs/dl/lib/python3.7/site-packages/ray/train/_internal/utils.py", line 183, in __getattr__
E                           actor_method = getattr(self.actor, item)
E                         [Previous line repeated 492 more times]
E                       RecursionError: maximum recursion depth exceeded while calling a Python object

../../../opt/anaconda3/envs/dl/lib/python3.7/site-packages/ray/_private/worker.py:2241: RayActorError
```
the suggested version seems having the error of `maximum recursion depth exceeded`.

**Comment by JiahaoYao** [comment]
and the output of the `TorchTrainer` is 
```python
[Result(metrics={'trial_id': '28897_00000', 'experiment_id': '93cd74990e27468ca5123dd840c51db6', 'date': '2022-07-25_10-49-44', 'timestamp': 1658771384, 'pid': 72773, 'hostname': 'jimmys-MBP.local.meter', 'node_ip': '127.0.0.1', 'config': {}, 'done': True}, checkpoint=<ray.air.checkpoint.Checkpoint object at 0x7f993ca2add0>, error=None, log_dir=PosixPath('/Users/jimmy/ray_results/TorchTrainer_2022-07-25_10-49-40/TorchTrainer_28897_00000_0_2022-07-25_10-49-41'), metrics_dataframe=None, best_checkpoints=[(<ray.air.checkpoint.Checkpoint object at 0x7f993ca31150>, {'trial_id': '28897_00000', 'experiment_id': '93cd74990e27468ca5123dd840c51db6', 'date': '2022-07-25_10-49-44', 'timestamp': 1658771384, 'pid': 72773, 'hostname': 'jimmys-MBP.local.meter', 'node_ip': '127.0.0.1', 'config': {}, 'done': True})])]
```

**Comment by amogkam** [comment]
@JiahaoYao can you provide more context for the comments that you posted?

**Comment by JiahaoYao** [comment]
i fixed these issue, i) move the trainer into the remote actor; 2) adding some burning time in order to avoid the same checkpoint name conflict. 

for the second point i raise, the current `TorchTrainer` can not passed the results/output to outside. The only way is to use `session.report` (because this only reports on the local rank o worker). Thus, the current workaround is to test inside the `train_func`, where we are able to access to all workers.  

------

In summary, i fixed that issue reported here and i just updated the test, and it passed locally. please review again after the ci passed.

File: python/ray/train/tests/test_torch_trainer.py [hunk 6]
```diff
@@ -111,6 +129,58 @@ def test_checkpoint_freq(ray_start_4_cpus):
...
+@pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2])
+def test_tune_torch_get_device_gpu(ray_2_node_4_gpu, num_gpus_per_worker):
+    """Tests if GPU ids are set correctly when running train concurrently in nested
+    actors (for example when used with Tune)."""
+    from ray.air.config import ScalingConfig
+    from ray.air import session
+    import time
+
+    num_samples = 2
+    num_workers = 2
```
Comment (amogkam) [inline python/ray/train/tests/test_torch_trainer.py:141]: This should be set to `2 / num_gpus_per_worker`?

Otherwise the placement can happen anywhere on the cluster and the test can be non-deterministic?

File: python/ray/train/tests/test_torch_trainer.py [hunk 7]
```diff
@@ -111,6 +129,58 @@ def test_checkpoint_freq(ray_start_4_cpus):
         trainer.fit()
 
 
+@pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2])
+def test_tune_torch_get_device_gpu(ray_2_node_4_gpu, num_gpus_per_worker):
```
Comment (amogkam) [inline python/ray/train/tests/test_torch_trainer.py:133]: Should we make this ray_2_node_2_gpu? Otherwise, if we have more GPUs in the cluster than workers, then placement can be non-deterministic?

Comment (JiahaoYao) [reply python/ray/train/tests/test_torch_trainer.py]: changed to `ray_2_node_2_gpu `

File: python/ray/train/tests/test_torch_trainer.py (hunk 6 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_torch_trainer.py]: thanks, make sense, done

**Review by JiahaoYao — COMMENTED** [review]
please review again? TYTY

File: python/ray/train/tests/test_torch_trainer.py [hunk 8]
```diff
@@ -111,6 +129,58 @@ def test_checkpoint_freq(ray_start_4_cpus):
         trainer.fit()
 
 
+@pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2])
+def test_tune_torch_get_device_gpu(ray_2_node_2_gpu, num_gpus_per_worker):
+    """Tests if GPU ids are set correctly when running train concurrently in nested
+    actors (for example when used with Tune)."""
```
Comment (amogkam) [inline python/ray/train/tests/test_torch_trainer.py:135]: Nit: Make docstring 1 line

File: python/ray/train/torch/train_loop_utils.py [hunk 9]
```diff
@@ -503,7 +503,10 @@ def get_device(self) -> torch.device:
                     device_id = cuda_visible_list.index(gpu_id)
                 else:
                     raise RuntimeError(
-                        f"CUDA_VISIBLE_DEVICES set incorrectly: {cuda_visible_str}"
+                        "CUDA_VISIBLE_DEVICES set incorrectly. "
+                        f"Got {cuda_visible_str}, expected to include {gpu_id}. "
+                        "Did you override `CUDA_VISIBLE_DEVICES`? "
```
Comment (amogkam) [inline python/ray/train/torch/train_loop_utils.py:508]:
```suggestion
                        "Did you override the `CUDA_VISIBLE_DEVICES`?  environment variable"
```

**Review by amogkam — APPROVED** [review]
Thanks @JiahaoYao, lgtm!

File: python/ray/train/tests/test_torch_trainer.py (hunk 8 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_torch_trainer.py:135]: small issue with lint

```
python/ray/train/torch/train_loop_utils.py:508:89: E501 line too long (92 > 88 characters)
```

Comment (JiahaoYao) [reply python/ray/train/tests/test_torch_trainer.py:135]: done

**Review by JiahaoYao — COMMENTED** [review]
merge the new change, and review again, TYTY

**Review by richardliaw — APPROVED** [review]
